File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/handlers/Outcomes.java
package com.azure.ai.vision.face.deviceattestation.handlers;

import java.util.Map;

import com.azure.ai.vision.face.deviceattestation.logging.AttestationLogger;
import com.azure.ai.vision.face.deviceattestation.services.ApiTelemetry;

/**
 * Builds {@link HandlerOutcome} values. {@link #fail} emits the failure
 * telemetry in the same step so the response code and telemetry code can never
 * drift.
 */
public final class Outcomes {

    /** Build a success (or otherwise non-failure) outcome. */
    public static HandlerOutcome ok(Object body, int status, String code, Object data) {
        boolean ok = status < 400;
        String message = body instanceof ErrorBody e ? e.message : null;
        return new HandlerOutcome(ok, code != null ? code : (ok ? "OK" : "ERROR"), status, body, message, data);
    }

    /** Build a 200 success outcome. */
    public static HandlerOutcome ok(Object body) {
        return ok(body, 200, null, null);
    }

    /** Build a FAILURE outcome AND emit its failure telemetry in one step. */
    public static HandlerOutcome fail(
            AttestationLogger logger,
            String route,
            int status,
            String code,
            String message,
            Map<String, Object> properties,
            String expiredAt,
            String validFrom) {
        ApiTelemetry.trackApiFail(logger, route, code, status, properties);
        return new HandlerOutcome(false, code, status, new ErrorBody(message, expiredAt, validFrom), message, null);
    }

    private Outcomes() {
    }
}

File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Handlers/Outcomes.cs
using Azure.AI.Vision.Face.DeviceAttestation.Logging;
using Azure.AI.Vision.Face.DeviceAttestation.Services;

namespace Azure.AI.Vision.Face.DeviceAttestation.Handlers;

/// <summary>
/// Builds <see cref="HandlerOutcome"/> values. <see cref="Fail"/> emits the
/// failure telemetry in the same step so the response code and telemetry code
/// can never drift.
/// </summary>
internal static class Outcomes
{
    /// <summary>Build a success (or otherwise non-failure) outcome.</summary>
    public static HandlerOutcome Ok(object body, int status = 200, string? code = null, object? data = null)
    {
        bool ok = status < 400;
        return new HandlerOutcome
        {
            Ok = ok,
            Code = code ?? (ok ? "OK" : "ERROR"),
            Status = status,
            Body = body,
            Message = body is ErrorBody e ? e.Message : null,
            Data = data,
        };
    }

    /// <summary>Build a FAILURE outcome AND emit its failure telemetry in one step.</summary>
    public static HandlerOutcome Fail(
        IAttestationLogger logger,
        string route,
        int status,
        string code,
        string message,
        IReadOnlyDictionary<string, object?>? properties = null,
        string? expiredAt = null,
        string? validFrom = null)
    {
        ApiTelemetry.TrackApiFail(logger, route, code, status, properties);
        return new HandlerOutcome
        {
            Ok = false,
            Code = code,
            Status = status,
            Message = message,
            Body = new ErrorBody { Message = message, ExpiredAt = expiredAt, ValidFrom = validFrom },
        };
    }
}
